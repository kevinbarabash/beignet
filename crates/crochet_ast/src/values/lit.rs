use std::fmt;
use swc_atoms::{Atom, JsWord};
use swc_common::source_map::DUMMY_SP;
use swc_ecma_ast;

use crate::types::{TLit, Type, TypeKind};
use crate::values::span::Span;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Num {
    pub span: Span,
    pub value: String,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Bool {
    pub span: Span,
    pub value: bool,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Str {
    pub span: Span,
    pub value: String,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Lit {
    // We store all of the values as strings since f64 doesn't
    // support the Eq trait because NaN and 0.1 + 0.2 != 0.3.
    Num(Num),
    Bool(Bool),
    Str(Str),
}

impl Lit {
    pub fn bool(value: bool, span: Span) -> Self {
        Lit::Bool(Bool { value, span })
    }

    pub fn str(value: String, span: Span) -> Self {
        Lit::Str(Str { value, span })
    }

    pub fn num(value: String, span: Span) -> Self {
        Lit::Num(Num { value, span })
    }

    pub fn span(&self) -> Span {
        match &self {
            Lit::Num(n) => n.span.to_owned(),
            Lit::Bool(b) => b.span.to_owned(),
            Lit::Str(s) => s.span.to_owned(),
        }
    }
}

impl fmt::Display for Lit {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Lit::Num(n) => write!(f, "{}", n.value),
            Lit::Bool(b) => write!(f, "{}", b.value),
            Lit::Str(s) => write!(f, "\"{}\"", s.value),
        }
    }
}

impl From<&Lit> for swc_ecma_ast::Expr {
    fn from(lit: &Lit) -> Self {
        match lit {
            Lit::Num(n) => {
                let lit = swc_ecma_ast::Lit::Num(swc_ecma_ast::Number {
                    span: DUMMY_SP,
                    // TODO: include the parsed value in the source AST node
                    value: n.value.parse().unwrap(),
                    // Use `None` value only for transformations to avoid recalculate
                    // characters in number literal
                    raw: Some(Atom::new(n.value.clone())),
                });
                swc_ecma_ast::Expr::Lit(lit)
            }
            Lit::Bool(b) => {
                let lit = swc_ecma_ast::Lit::Bool(swc_ecma_ast::Bool {
                    span: DUMMY_SP,
                    value: b.value,
                });
                swc_ecma_ast::Expr::Lit(lit)
            }
            Lit::Str(s) => {
                let lit = swc_ecma_ast::Lit::Str(swc_ecma_ast::Str {
                    span: DUMMY_SP,
                    value: JsWord::from(s.value.clone()),
                    // Use `None` value only for transformations to avoid recalculate escaped
                    // characters in strings
                    raw: None,
                });
                swc_ecma_ast::Expr::Lit(lit)
            }
        }
    }
}

impl From<Lit> for Type {
    fn from(lit: Lit) -> Self {
        Type {
            kind: TypeKind::Lit(match lit {
                Lit::Num(n) => TLit::Num(n.value),
                Lit::Bool(b) => TLit::Bool(b.value),
                Lit::Str(s) => TLit::Str(s.value),
            }),
            provenance: None,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn boolean() {
        let t = Lit::bool(true, 0..4);
        let f = Lit::bool(false, 0..5);
        assert_eq!(format!("{}", t), "true");
        assert_eq!(format!("{}", f), "false");
        assert_eq!(t.span(), 0..4);
    }

    #[test]
    fn number() {
        let num = Lit::num(String::from("1.23"), 0..4);
        assert_eq!(format!("{}", num), "1.23");
        assert_eq!(num.span(), 0..4);
    }

    #[test]
    fn string() {
        let s = Lit::str(String::from("hello"), 0..5);
        assert_eq!(format!("{}", s), "\"hello\"");
        assert_eq!(s.span(), 0..5);
    }
}
