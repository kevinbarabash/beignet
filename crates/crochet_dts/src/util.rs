use std::collections::HashMap;

use swc_ecma_ast::*;

use crochet_ast::types::{
    self as types, TCallable, TFnParam, TGeneric, TIndexAccess, TObjElem, TObject, TVar, Type,
    TypeKind,
};
use crochet_infer::{get_type_params, set_type_params, Context, Subst, Substitutable};

use crate::parse_dts::infer_ts_type_ann;

// TODO: rename this replace_refs
pub fn replace_aliases(
    t: &Type,
    type_param_decl: &TsTypeParamDecl,
    ctx: &Context,
) -> Result<Type, String> {
    let mut type_params: Vec<TVar> = vec![];
    let type_param_map: HashMap<String, TVar> = type_param_decl
        .params
        .iter()
        .map(|tp| {
            // NOTE: We can't use .map() here because infer_ts_type_ann
            // returns a Result.
            let constraint = match &tp.constraint {
                Some(constraint) => {
                    let t = infer_ts_type_ann(constraint, ctx)?;
                    Some(Box::from(t))
                }
                None => None,
            };
            let tv = TVar {
                id: ctx.fresh_id(),
                constraint,
            };
            type_params.push(tv.clone());
            Ok((tp.name.sym.to_string(), tv))
        })
        .collect::<Result<HashMap<String, TVar>, String>>()?;

    let t = set_type_params(t, &type_params);
    Ok(replace_aliases_rec(&t, &type_param_map))
}

// TODO: rename this replace_refs_rec
fn replace_aliases_rec(t: &Type, map: &HashMap<String, TVar>) -> Type {
    match &t.kind {
        TypeKind::Generic(TGeneric { t, type_params }) => {
            // TODO: create a new `map` that adds in `type_params`
            Type {
                kind: TypeKind::Generic(TGeneric {
                    t: Box::from(replace_aliases_rec(t, map)),
                    type_params: type_params.to_owned(),
                }),
            }
        }
        TypeKind::Var(_) => t.to_owned(),
        TypeKind::App(types::TApp { args, ret }) => Type {
            kind: TypeKind::App(types::TApp {
                args: args.iter().map(|t| replace_aliases_rec(t, map)).collect(),
                ret: Box::from(replace_aliases_rec(ret, map)),
            }),
        },
        TypeKind::Lam(types::TLam { params, ret }) => Type {
            kind: TypeKind::Lam(types::TLam {
                params: params
                    .iter()
                    .map(|param| TFnParam {
                        t: replace_aliases_rec(&param.t, map),
                        ..param.to_owned()
                    })
                    .collect(),
                ret: Box::from(replace_aliases_rec(ret, map)),
            }),
        },
        TypeKind::Lit(_) => t.to_owned(),
        TypeKind::Keyword(_) => t.to_owned(),
        TypeKind::Union(types) => Type {
            kind: TypeKind::Union(types.iter().map(|t| replace_aliases_rec(t, map)).collect()),
        },
        TypeKind::Intersection(types) => Type {
            kind: TypeKind::Intersection(
                types.iter().map(|t| replace_aliases_rec(t, map)).collect(),
            ),
        },
        TypeKind::Object(obj) => {
            let elems: Vec<TObjElem> = obj
                .elems
                .iter()
                .map(|elem| match elem {
                    TObjElem::Call(lam) => {
                        let params: Vec<TFnParam> = lam
                            .params
                            .iter()
                            .map(|t| TFnParam {
                                t: replace_aliases_rec(&t.t, map),
                                ..t.to_owned()
                            })
                            .collect();
                        let ret = replace_aliases_rec(lam.ret.as_ref(), map);

                        TObjElem::Call(TCallable {
                            params,
                            ret: Box::from(ret),
                            type_params: lam.type_params.to_owned(),
                        })
                    }
                    TObjElem::Constructor(lam) => {
                        let params: Vec<TFnParam> = lam
                            .params
                            .iter()
                            .map(|t| TFnParam {
                                t: replace_aliases_rec(&t.t, map),
                                ..t.to_owned()
                            })
                            .collect();
                        let ret = replace_aliases_rec(lam.ret.as_ref(), map);

                        TObjElem::Constructor(TCallable {
                            params,
                            ret: Box::from(ret),
                            type_params: lam.type_params.to_owned(),
                        })
                    }
                    TObjElem::Index(index) => {
                        let t = replace_aliases_rec(&index.t, map);
                        TObjElem::Index(types::TIndex {
                            t,
                            ..index.to_owned()
                        })
                    }
                    TObjElem::Prop(prop) => {
                        println!("replacing prop.t");
                        println!("prop.t = {:#}", prop.t);
                        let t = replace_aliases_rec(&prop.t, map);
                        TObjElem::Prop(types::TProp {
                            t,
                            ..prop.to_owned()
                        })
                    }
                })
                .collect();
            Type {
                kind: TypeKind::Object(TObject { elems }),
            }
        }
        TypeKind::Ref(alias) => match map.get(&alias.name) {
            Some(tv) => Type {
                kind: TypeKind::Var(tv.to_owned()),
            },
            None => t.to_owned(),
        },
        TypeKind::Tuple(types) => Type {
            kind: TypeKind::Tuple(types.iter().map(|t| replace_aliases_rec(t, map)).collect()),
        },
        TypeKind::Array(t) => Type {
            kind: TypeKind::Array(Box::from(replace_aliases_rec(t, map))),
        },
        TypeKind::Rest(t) => Type {
            kind: TypeKind::Rest(Box::from(replace_aliases_rec(t, map))),
        },
        TypeKind::This => Type {
            kind: TypeKind::This,
        },
        TypeKind::KeyOf(t) => Type {
            kind: TypeKind::KeyOf(Box::from(replace_aliases_rec(t, map))),
        },
        TypeKind::IndexAccess(TIndexAccess { object, index }) => Type {
            kind: TypeKind::IndexAccess(TIndexAccess {
                object: Box::from(replace_aliases_rec(object, map)),
                index: Box::from(replace_aliases_rec(index, map)),
            }),
        },
        TypeKind::Mutable(t) => Type {
            kind: TypeKind::Mutable(Box::from(replace_aliases_rec(t, map))),
        },
    }
}

pub fn merge_types(t1: &Type, t2: &Type) -> Type {
    let tp1 = get_type_params(t1);
    let tp2 = get_type_params(t2);

    if tp1.len() != tp2.len() {
        panic!("Mismatch in type param count when attempting to merge type");
    }

    // Creates a mapping from type params in t2 to those in t1
    let subs: Subst = tp2
        .into_iter()
        .map(|tv| tv.id.to_owned())
        .zip(tp1.iter().map(|tv| Type {
            kind: TypeKind::Var(tv.to_owned()),
        }))
        .collect();

    // Unwrap qualified types since we return a qualified
    // type if there are any type qualifiers.
    let t1 = if let TypeKind::Generic(TGeneric { t, .. }) = &t1.kind {
        t
    } else {
        t1
    };
    let t2 = if let TypeKind::Generic(TGeneric { t, .. }) = &t2.kind {
        t
    } else {
        t2
    };

    // Updates type variables for type params to match t1
    let t2 = t2.apply(&subs);

    let type_params = tp1;
    let t = match (&t1.kind, &t2.kind) {
        (TypeKind::Object(obj1), TypeKind::Object(obj2)) => {
            let elems: Vec<_> = obj1
                .elems
                .iter()
                .cloned()
                .chain(obj2.elems.iter().cloned())
                .collect();

            Type {
                kind: TypeKind::Object(TObject { elems }),
            }
        }
        (_, _) => todo!(),
    };

    if type_params.is_empty() {
        t
    } else {
        Type {
            kind: TypeKind::Generic(TGeneric {
                t: Box::from(t),
                type_params,
            }),
        }
    }
}
