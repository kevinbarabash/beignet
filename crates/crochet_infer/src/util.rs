use defaultmap::*;

use std::collections::{HashMap, HashSet};
use std::iter::Iterator;

use super::context::{Context, Env};
use super::substitutable::{Subst, Substitutable};
use super::types::*;

pub fn normalize(sc: &Scheme, ctx: &Context) -> Scheme {
    let body = &sc.ty;
    let keys = body.ftv();
    let mut keys: Vec<_> = keys.iter().cloned().collect();
    keys.sort_unstable();
    let mapping: HashMap<i32, Type> = keys
        .iter()
        .enumerate()
        .map(|(index, key)| (key.to_owned(), Type::Var(index as i32)))
        .collect();

    // TODO: add norm_type as a method on Type, Vec<Type>, etc. similar to what we do for Substitutable
    fn norm_type(ty: &Type, mapping: &HashMap<i32, Type>, ctx: &Context) -> Type {
        match &ty {
            Type::Var(id) => mapping.get(id).unwrap().to_owned(),
            Type::App(app) => {
                let args: Vec<_> = app
                    .args
                    .iter()
                    .map(|arg| norm_type(arg, mapping, ctx))
                    .collect();
                let ret = Box::from(norm_type(&app.ret, mapping, ctx));
                Type::App(AppType { args, ret })
            }
            Type::Lam(lam) => {
                let params: Vec<_> = lam
                    .params
                    .iter()
                    .map(|param| TFnParam {
                        ty: norm_type(&param.ty, mapping, ctx),
                        ..param.to_owned()
                    })
                    .collect();
                let ret = Box::from(norm_type(&lam.ret, mapping, ctx));
                Type::Lam(LamType { params, ret })
            }
            Type::Wildcard => ty.to_owned(),
            Type::Prim(_) => ty.to_owned(),
            Type::Lit(_) => ty.to_owned(),
            Type::Union(types) => {
                // TODO: update union_types from constraint_solver.rs to handle
                // any number of types instead of just two and then call it here.
                let types = types.iter().map(|ty| norm_type(ty, mapping, ctx)).collect();
                Type::Union(types)
            }
            Type::Intersection(types) => {
                // TODO: update intersection_types from constraint_solver.rs to handle
                // any number of types instead of just two and then call it here.
                let types: Vec<_> = types.iter().map(|ty| norm_type(ty, mapping, ctx)).collect();
                simplify_intersection(&types, ctx)
            }
            Type::Object(props) => {
                let props = props
                    .iter()
                    .map(|prop| TProp {
                        name: prop.name.clone(),
                        optional: prop.optional,
                        mutable: prop.mutable,
                        // NOTE: we don't use prop.get_type(ctx) here because we're tracking
                        // the optionality of the property in the TProp that's returned.
                        ty: norm_type(&prop.ty, mapping, ctx),
                    })
                    .collect();
                Type::Object(props)
            }
            Type::Alias(AliasType { name, type_params }) => {
                let type_params = type_params.clone().map(|params| {
                    params
                        .iter()
                        .map(|ty| norm_type(ty, mapping, ctx))
                        .collect()
                });
                Type::Alias(AliasType {
                    name: name.to_owned(),
                    type_params,
                })
            }
            Type::Tuple(types) => {
                let types = types.iter().map(|ty| norm_type(ty, mapping, ctx)).collect();
                Type::Tuple(types)
            }
            Type::Array(t) => Type::Array(Box::from(norm_type(t, mapping, ctx))),
            Type::Rest(arg) => Type::Rest(Box::from(norm_type(arg, mapping, ctx))),
        }
    }

    Scheme {
        qualifiers: (0..keys.len()).map(|x| x as i32).collect(),
        ty: norm_type(body, &mapping, ctx),
    }
}

pub fn generalize(env: &Env, ty: &Type) -> Scheme {
    // ftv() returns a Set which is not ordered
    // TODO: switch to an ordered set
    let mut qualifiers: Vec<_> = ty.ftv().difference(&env.ftv()).cloned().collect();
    qualifiers.sort_unstable();
    Scheme {
        qualifiers,
        ty: ty.clone(),
    }
}

// TODO: make this recursive
// TODO: handle optional properties correctly
// Maybe we can have a function that will canonicalize objects by converting
// `x: T | undefined` to `x?: T`
pub fn simplify_intersection(in_types: &[Type], ctx: &Context) -> Type {
    let obj_types: Vec<_> = in_types
        .iter()
        .filter_map(|ty| match &ty {
            Type::Object(props) => Some(props),
            _ => None,
        })
        .collect();

    // The use of HashSet<Type> here is to avoid duplicate types
    let mut props_map: DefaultHashMap<String, HashSet<Type>> = defaulthashmap!();
    for props in obj_types {
        for prop in props {
            props_map[prop.name.clone()].insert(prop.ty.clone());
        }
    }

    let mut props: Vec<TProp> = props_map
        .iter()
        .map(|(name, types)| {
            let types: Vec<_> = types.iter().cloned().collect();
            let ty: Type = if types.len() == 1 {
                types[0].clone()
            } else {
                ctx.intersection(types)
            };
            TProp {
                name: name.to_owned(),
                // TODO: determine this field from all of the TProps with
                // the same name.  This should only be optional if all of
                // the TProps with the current name are optional.
                optional: false,
                mutable: false,
                ty,
            }
        })
        .collect();
    props.sort_by_key(|prop| prop.name.clone()); // ensure a stable order

    let mut not_obj_types: Vec<_> = in_types
        .iter()
        .filter(|ty| !matches!(ty, Type::Object(_)))
        .cloned()
        .collect();

    let mut out_types = vec![];
    out_types.append(&mut not_obj_types);
    if !props.is_empty() {
        out_types.push(ctx.object(props));
    }
    // TODO: figure out a consistent way to sort types
    // out_types.sort_by_key(|ty| ty.id); // ensure a stable order

    if out_types.len() == 1 {
        out_types[0].clone()
    } else {
        ctx.intersection(out_types)
    }
}

fn flatten_types(ty: &Type) -> Vec<Type> {
    match &ty {
        Type::Union(types) => types.iter().flat_map(flatten_types).collect(),
        _ => vec![ty.to_owned()],
    }
}

pub fn union_types(t1: &Type, t2: &Type, ctx: &Context) -> Type {
    union_many_types(&[t1.to_owned(), t2.to_owned()], ctx)
}

pub fn union_many_types(ts: &[Type], ctx: &Context) -> Type {
    let types: Vec<_> = ts.iter().flat_map(flatten_types).collect();

    let types_set: HashSet<_> = types.iter().cloned().collect();

    let prim_types: HashSet<_> = types_set
        .iter()
        .cloned()
        .filter(|ty| matches!(ty, Type::Prim(_)))
        .collect();
    let lit_types: HashSet<_> = types_set
        .iter()
        .cloned()
        .filter(|ty| match &ty {
            // Primitive types subsume corresponding literal types
            Type::Lit(lit) => match lit {
                Lit::Num(_) => !prim_types.contains(&ctx.prim(Primitive::Num)),
                Lit::Bool(_) => !prim_types.contains(&ctx.prim(Primitive::Bool)),
                Lit::Str(_) => !prim_types.contains(&ctx.prim(Primitive::Str)),
                Lit::Null => !prim_types.contains(&ctx.prim(Primitive::Null)),
                Lit::Undefined => !prim_types.contains(&ctx.prim(Primitive::Undefined)),
            },
            _ => false,
        })
        .collect();
    let rest_types: HashSet<_> = types_set
        .iter()
        .cloned()
        .filter(|ty| !matches!(ty, Type::Prim(_) | Type::Lit(_)))
        .collect();

    let types: Vec<_> = prim_types
        .iter()
        .chain(lit_types.iter())
        .chain(rest_types.iter())
        .cloned()
        .collect();

    if types.len() > 1 {
        ctx.union(types)
    } else {
        types[0].clone()
    }
}

pub fn compose_subs(s2: &Subst, s1: &Subst) -> Subst {
    let mut result: Subst = s1.iter().map(|(id, tv)| (*id, tv.apply(s2))).collect();
    result.extend(s2.to_owned());
    result
}

// subs are composed from left to right with ones to the right
// being applied to all of the ones to the left.
pub fn compose_many_subs(subs: &[Subst]) -> Subst {
    subs.iter()
        .fold(Subst::new(), |accum, next| compose_subs(&accum, next))
}

// If are multiple entries for the same type variable, this function merges
// them into a union type (simplifying the type if possible).
fn compose_subs_with_context(s1: &Subst, s2: &Subst, ctx: &Context) -> Subst {
    let mut result: Subst = s2.iter().map(|(i, t)| (*i, t.apply(s1))).collect();
    for (i, t) in s1 {
        match result.get(i) {
            Some(t1) => {
                let t = union_types(t, t1, ctx);
                result.insert(*i, t)
            }
            None => result.insert(*i, t.to_owned()),
        };
    }
    result
}

pub fn compose_many_subs_with_context(subs: &[Subst], ctx: &Context) -> Subst {
    subs.iter().fold(Subst::new(), |accum, next| {
        compose_subs_with_context(&accum, next, ctx)
    })
}
