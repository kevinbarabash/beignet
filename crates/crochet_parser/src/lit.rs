use chumsky::prelude::*;
use crochet_ast::*;
use unescape::unescape;

pub fn string_parser() -> BoxedParser<'static, char, Lit, Simple<char>> {
    let escape = just('\\').ignore_then(
        just('\\')
            .or(just('/'))
            .or(just('"'))
            .or(just('b').to('\x08'))
            .or(just('f').to('\x0C'))
            .or(just('n').to('\n'))
            .or(just('r').to('\r'))
            .or(just('t').to('\t'))
            .or(just('u').ignore_then(
                filter(|c: &char| c.is_ascii_hexdigit())
                    .repeated()
                    .exactly(4)
                    .collect::<String>()
                    .validate(|digits, span, emit| {
                        char::from_u32(u32::from_str_radix(&digits, 16).unwrap()).unwrap_or_else(
                            || {
                                emit(Simple::custom(span, "invalid unicode character"));
                                '\u{FFFD}' // unicode replacement character
                            },
                        )
                    }),
            )),
    );

    let string = just("\"")
        .ignore_then(filter(|c| *c != '\\' && *c != '"').or(escape).repeated())
        .then_ignore(just("\""))
        .collect::<String>()
        .map_with_span(|raw, span| {
            // unescape needs to know whether the string is contained in single quotes
            // or double quotes so that it can unescape quote characters correctly.
            let cooked = unescape(&raw).unwrap();
            Lit::str(cooked, span)
        });

    string.boxed()
}
