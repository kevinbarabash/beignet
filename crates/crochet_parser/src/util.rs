use chumsky::prelude::*;

pub fn comment_parser() -> BoxedParser<'static, char, &'static str, Simple<char>> {
    let comment = just("//")
        .then_ignore(take_until(just('\n')))
        .padded()
        .labelled("comment");

    comment.boxed()
}

pub fn just_with_padding(input: &str) -> BoxedParser<'_, char, &str, Simple<char>> {
    just(input)
        .padded()
        .padded_by(comment_parser().repeated())
        .boxed()
}
