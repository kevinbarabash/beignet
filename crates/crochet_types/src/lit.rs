use std::fmt;
use std::hash::Hash;

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub enum Lit {
    // We store all of the values as strings since f64 doesn't
    // support the Eq trait because NaN and 0.1 + 0.2 != 0.3.
    Num(String),
    Bool(bool),
    Str(String),
    Null,
    Undefined,
}

impl fmt::Display for Lit {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Lit::Num(n) => write!(f, "{}", n),
            Lit::Bool(b) => write!(f, "{}", b),
            Lit::Str(s) => write!(f, "\"{}\"", s),
            Lit::Null => write!(f, "null"),
            Lit::Undefined => write!(f, "undefined"),
        }
    }
}
