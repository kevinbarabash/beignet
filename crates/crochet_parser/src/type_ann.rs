use chumsky::prelude::*;
use crochet_ast::*;

use crate::lit::{boolean_parser, number_parser, string_parser};
use crate::util::just_with_padding;

use super::type_params::type_params;

pub fn type_ann_parser() -> BoxedParser<'static, char, TypeAnn, Simple<char>> {
    let prim = choice((
        just("number").to(Primitive::Num),
        just("string").to(Primitive::Str),
        just("boolean").to(Primitive::Bool),
    ))
    .map_with_span(|prim, span| TypeAnn::Prim(PrimType { span, prim }))
    .padded();

    let keyword = choice((
        just("null").to(Keyword::Null),
        just("symbol").to(Keyword::Symbol),
        just("undefined").to(Keyword::Undefined),
    ))
    .padded()
    .map_with_span(|keyword, span| TypeAnn::Keyword(KeywordType { keyword, span }));

    let parser = recursive(|type_ann| {
        let type_args = type_ann
            .clone()
            .separated_by(just_with_padding(","))
            .allow_trailing()
            .delimited_by(just_with_padding("<"), just_with_padding(">"));

        let type_ref =
            text::ident()
                .then(type_args.or_not())
                .map_with_span(|(name, type_params), span| {
                    TypeAnn::TypeRef(TypeRef {
                        span,
                        name,
                        type_params,
                    })
                });

        let prop = text::ident()
            .then(just("?").or_not())
            .then_ignore(just_with_padding(":"))
            .then(type_ann.clone())
            .map_with_span(|((name, optional), type_ann), span: Span| TProp {
                span,
                name,
                optional: optional.is_some(),
                // TODO: properties are immutable by default, add a `mut` keyword
                // to make them mutable
                mutable: false,
                type_ann: Box::from(type_ann),
            });

        let obj = prop
            .separated_by(just_with_padding(","))
            .allow_trailing()
            .delimited_by(just_with_padding("{"), just_with_padding("}"))
            .map_with_span(|props, span: Span| TypeAnn::Object(ObjectType { span, props }));

        let tuple = type_ann
            .clone()
            .separated_by(just_with_padding(","))
            .allow_trailing()
            .delimited_by(just_with_padding("["), just_with_padding("]"))
            .map_with_span(|types, span: Span| TypeAnn::Tuple(TupleType { span, types }));

        let atom = choice((
            boolean_parser().map(TypeAnn::Lit),
            number_parser().map(TypeAnn::Lit),
            string_parser().map(TypeAnn::Lit),
            prim,
            keyword,
            obj,
            tuple,
            type_ref,
            type_ann
                .clone()
                .delimited_by(just_with_padding("("), just_with_padding(")")),
        ));

        let atom_with_suffix =
            atom.clone()
                .then(just_with_padding("[]").repeated())
                .foldl(|accum, _| {
                    TypeAnn::Array(ArrayType {
                        span: 0..0, // FIXME
                        elem_type: Box::from(accum),
                    })
                });

        // We have to use `atom` here instead of `type_ann` to avoid a stack
        // overflow.
        let intersection = atom_with_suffix
            .clone()
            .separated_by(just_with_padding("&"))
            .map_with_span(|types, span| match types.len() {
                1 => types[0].clone(),
                _ => TypeAnn::Intersection(IntersectionType { span, types }),
            });

        let union = just_with_padding("|")
            .or_not()
            .ignore_then(intersection.separated_by(just_with_padding("|")))
            .map_with_span(|types, span| match types.len() {
                1 => types[0].clone(),
                _ => TypeAnn::Union(UnionType { span, types }),
            });

        let ident = text::ident().map_with_span(|name, span: Span| Ident { span, name });

        let ident_param = ident
            .then(just_with_padding("?").or_not())
            .then_ignore(just_with_padding(":"))
            .then(type_ann.clone())
            .map_with_span(
                |((ident, question_mark), type_ann), span: Span| TypeAnnFnParam {
                    pat: EFnParamPat::Ident(EFnParamBindingIdent { span, id: ident }),
                    type_ann,
                    optional: question_mark.is_some(),
                },
            );

        let rest_param = just_with_padding("...")
            .ignore_then(ident.map_with_span(|id, span: Span| {
                EFnParamPat::Ident(EFnParamBindingIdent { span, id })
            }))
            .then_ignore(just_with_padding(":"))
            .then(type_ann.clone())
            .map_with_span(|(ident, type_ann), span: Span| TypeAnnFnParam {
                pat: EFnParamPat::Rest(EFnParamRestPat {
                    span,
                    arg: Box::from(ident),
                }),
                type_ann,
                optional: false,
            });

        let lam_params = choice((ident_param, rest_param))
            .separated_by(just_with_padding(","))
            .allow_trailing()
            .delimited_by(just_with_padding("("), just_with_padding(")"));

        let lam = type_params(type_ann.clone().boxed())
            .or_not()
            .then(lam_params)
            .then_ignore(just_with_padding("=>"))
            .then(type_ann.clone())
            .map_with_span(|((type_params, params), ret), span| {
                TypeAnn::Lam(LamType {
                    span,
                    params,
                    ret: Box::from(ret),
                    type_params,
                })
            });

        choice((
            // lambda types have higher precedence than union types so that
            // `(a: A, b: B) => C | D` parse as lambda type with a return type that
            // happens to be a union.
            lam, union, atom,
        ))
    });

    parser.boxed()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn parse_type(input: &str) -> TypeAnn {
        type_ann_parser().then_ignore(end()).parse(input).unwrap()
    }

    #[test]
    fn fn_type_annotations() {
        insta::assert_debug_snapshot!(parse_type("(a: number, b: string) => boolean"));
        insta::assert_debug_snapshot!(parse_type("(a: A) => B & C | D"));
        insta::assert_debug_snapshot!(parse_type("(a: A, b: B) => C & D"));
        insta::assert_debug_snapshot!(parse_type("(a: A, b: B) => C | D"));
        insta::assert_debug_snapshot!(parse_type("((a: A, b: B) => C) | D"));
    }

    #[test]
    fn union_and_intersection_type_annotations() {
        insta::assert_debug_snapshot!(parse_type("A & B & C"));
        insta::assert_debug_snapshot!(parse_type("A & B | C & D"));
        insta::assert_debug_snapshot!(parse_type("(A | B) & (C | D)"));
    }

    #[test]
    fn literal_type_annotations() {
        insta::assert_debug_snapshot!(parse_type(r#""hello""#));
        insta::assert_debug_snapshot!(parse_type("true"));
        insta::assert_debug_snapshot!(parse_type("false"));
        insta::assert_debug_snapshot!(parse_type("1.23"));
        insta::assert_debug_snapshot!(parse_type("5"));
    }

    #[test]
    fn misc_type_annotations() {
        insta::assert_debug_snapshot!(parse_type("Promise<number>"));
        insta::assert_debug_snapshot!(parse_type("{x: number, y: number}"));
    }
}
