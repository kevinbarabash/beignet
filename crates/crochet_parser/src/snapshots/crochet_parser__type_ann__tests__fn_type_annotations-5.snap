---
source: crates/crochet_parser/src/type_ann.rs
expression: "parse_type(\"((a: A, b: B) => C) | D\")"
---
Union(
    UnionType {
        span: 0..23,
        types: [
            Lam(
                LamType {
                    span: 1..18,
                    params: [
                        Ident(
                            BindingIdent {
                                span: 2..6,
                                id: Ident {
                                    span: 2..3,
                                    name: "a",
                                },
                                type_ann: Some(
                                    TypeRef(
                                        TypeRef {
                                            span: 5..6,
                                            name: "A",
                                            type_params: None,
                                        },
                                    ),
                                ),
                            },
                        ),
                        Ident(
                            BindingIdent {
                                span: 8..12,
                                id: Ident {
                                    span: 8..9,
                                    name: "b",
                                },
                                type_ann: Some(
                                    TypeRef(
                                        TypeRef {
                                            span: 11..12,
                                            name: "B",
                                            type_params: None,
                                        },
                                    ),
                                ),
                            },
                        ),
                    ],
                    ret: TypeRef(
                        TypeRef {
                            span: 17..18,
                            name: "C",
                            type_params: None,
                        },
                    ),
                    type_params: None,
                },
            ),
            TypeRef(
                TypeRef {
                    span: 22..23,
                    name: "D",
                    type_params: None,
                },
            ),
        ],
    },
)
